app/project_type/git.py

```diff
@@ -149,25 +150,38 @@ def _fetch_project(self):
 
         # Must add the --update-head-ok in the scenario that the current branch of the working directory
         # is equal to self._branch, otherwise the git fetch will exit with a non-zero exit code.
-        # Must specify the colon in 'branch:branch' so that the branch will be created locally. This is
-        # important because it allows the slave hosts to do a git fetch from the master for this branch.
-        fetch_command = 'git fetch --update-head-ok {0} {1}:{1}'.format(self._remote, self._branch)
+        fetch_command = 'git fetch --update-head-ok {} {}'.format(self._remote, self._branch)
         self._git_remote_command_executor.execute(fetch_command, cwd=self._repo_directory)
 
-        commit_hash = self._hash or 'FETCH_HEAD'
+        # Validate and convert the user-specified hash/refspec to a full git hash
+        self._hash = self._execute_in_repo_and_raise_on_failure(
+            'git rev-parse {}'.format(self._hash),
+            'Could not rev-parse "{}" to a commit hash.'.format(self._hash)
+        ).strip()
+
+        # Save this hash as a local ref. Named local refs are necessary for slaves to fetch correctly from the master.
+        local_ref = 'refs/clusterrunner/' + self._hash
+        update_ref_command = 'git update-ref {} {}'.format(local_ref, self._hash)
+        self._execute_in_repo_and_raise_on_failure(update_ref_command, 'Could not update local ref.')
```

Comment:
Maybe include the local_ref and or self._hash in the error message? I'm sorry, you may have to go over 100 lines =P